File: bfs/cpp/bfs.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <queue>
#include <limits>

struct edge 
{
    int destination;
    int weight;
};

class graph 
{
public:
    explicit graph(int number_of_nodes) : adjacency_list_(number_of_nodes) {}

    void add_edge(int source, int destination, int weight) 
    {
        adjacency_list_[source].push_back({ destination, weight });
    }

    std::pair<std::vector<int>, int> shortest_path(int start, int end) const
    {
        std::vector<int> distances(adjacency_list_.size(), std::numeric_limits<int>::max());
        std::vector<int> previous(adjacency_list_.size(), -1);
        std::vector<bool> visited(adjacency_list_.size(), false);

        distances[start] = 0;
		
        std::queue<int> queue;
        queue.push(start);
        visited[start] = true;

        while (!queue.empty()) 
        {
            int current = queue.front();
            queue.pop();

            for (const edge& bfs_edge : adjacency_list_[current]) 
            {
                int neighbour = bfs_edge.destination;
                int weight = bfs_edge.weight;

                if (!visited[neighbour]) 
                {
                    visited[neighbour] = true;
                    queue.push(neighbour);
                }

                if (distances[current] + weight < distances[neighbour]) 
                {
                    distances[neighbour] = distances[current] + weight;
                    previous[neighbour] = current;
                }
            }
        }

        std::vector<int> path;
        int current = end;
        while (current != -1) 
        {
            path.push_back(current);
            current = previous[current];
        }

        std::reverse(path.begin(), path.end());
        return { path, distances[end] };
    }

private:
    std::vector<std::vector<edge>> adjacency_list_;
};

int main() 
{
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);

    std::string filename = "bfs_input.txt";
    std::ifstream file(filename);
    
    if (!file) 
    {
        std::cerr << "Failed to open the file." << std::endl;
        return 1;
    }

    int number_of_nodes, number_of_edges;
    file >> number_of_nodes >> number_of_edges;

    graph bfs_graph(number_of_nodes);

    for (int i = 0; i < number_of_edges; ++i) 
    {
        int source, destination, weight;
        file >> source >> destination >> weight;
        bfs_graph.add_edge(source, destination, weight);
    }

    int start, end;
    std::cout << "Enter starting node: ";
    std::cin >> start;
    std::cout << "Enter end node: ";
    std::cin >> end;

    std::pair<std::vector<int>, int> result = bfs_graph.shortest_path(start, end);
    std::vector<int> path = result.first;
    int shortest_distance = result.second;

    if (path[0] != start) 
    {
        std::cout << "No path exists between the specified nodes." << std::endl;
        return 0;
    }

    std::cout << "Shortest path: ";
    for (int node : path) 
    {
        std::cout << node << " ";
    }

    std::cout << std::endl << "Shortest distance: " << shortest_distance << std::endl;

    return 0;
}

File: bfs/python/bfs.py
from collections import deque

class Edge:
    def __init__(self, destination, weight):
        self.destination = destination
        self.weight = weight

class Graph:
    def __init__(self, number_of_vertices):
        self.adjacency_list = [[] for _ in range(number_of_vertices)]

    def AddEdge(self, source, destination, weight):
        self.adjacency_list[source].append(Edge(destination, weight))

    def ShortestPath(self, start, end):
        number_of_vertices = len(self.adjacency_list)
        distances = [float('inf')] * number_of_vertices
        previous = [-1] * number_of_vertices
        visited = [False] * number_of_vertices

        distances[start] = 0

        queue = deque()
        queue.append(start)
        visited[start] = True

        while queue:
            current = queue.popleft()

            for edge in self.adjacency_list[current]:
                neighbour = edge.destination
                weight = edge.weight

                if not visited[neighbour]:
                    visited[neighbour] = True
                    queue.append(neighbour)

                if distances[current] + weight < distances[neighbour]:
                    distances[neighbour] = distances[current] + weight
                    previous[neighbour] = current

        path = []
        current = end
        while current != -1:
            path.append(current)
            current = previous[current]

        path.reverse()
        return path, distances[end]

if __name__ == "__main__":
    filename = "bfs_input.txt"
    with open(filename, 'r') as file:
        number_of_vertices, number_of_edges = map(int, file.readline().split())

        graph = Graph(number_of_vertices)

        for _ in range(number_of_edges):
            source, destination, weight = map(int, file.readline().split())
            graph.AddEdge(source, destination, weight)

    start = int(input("Enter the starting node: "))
    end = int(input("Enter the end node: "))

    path, shortest_distance = graph.ShortestPath(start, end)

    if path[0] != start:
        print("No path exists between the specified nodes.")
    else:
        print("Shortest path:", " ".join(str(node) for node in path))
        print("Shortest distance:", shortest_distance)
